_posts/2024-01-11-gsoc2023_marinegor.md

```diff
@@ -0,0 +1,169 @@
+
+
+## A short description of the goals of the project.
+One sentence: introduce parallel execution of analysis runs in [MDAnalysis](https://github.com/MDAnalysis/mdanalysis) library. Somewhat good introduction I also gave [here](https://marinegor.github.io/posts/2023/05/gsoc-proposal/) when writing a proposal for the project.
+
+In more technical details, MDAnalysis library (as of v2.6.0) contains around 30 different subclasses that can perform various molecular trajectory analysis tasks, like calculating RMSD, RMSF, various contacts, density analysis, and more advanced tasks. 24 of these subclasses are children of `AnalysisBase` class. This base class is written in a way that allows subclass authors care about implementing only 3 methods:
+
+ 1. `_prepare()` -- how to initialize attributes for the analysis
+ 2. `_single_frame()` -- how to do analysis of a single molecular trajectory frame
+ 3. `_conclude()` -- how to transform intermediate results into final ones
+
+With only these 3 methods implemented, by inheritance subclasses get `run()` method that will take care of reading the trajectory, storing the results, logging, etc.
+
+The main goal was to re-write `AnalysisBase.run()` so that it can run in parallel on multiple processes, but make these changes invisible from the subclasses, i.e. not re-write any of their code.
+
+## What you did.
+Altogether, the `AnalysisBase.run()` has changed by addition of the following methods:
+
+ - `_setup_computation_groups()`: split frames into multiple parts for separate analysis
+ - `_compute()`: run `_single_frame` on a list of frames, but without running `_conclude`
+ - `_get_aggregator()`: get an object to aggregate the run results with, making them compatible with subsequent `_conclude`
+ - class property `available_backends()`: get list of `str` values that describe available backends for a given subclass
+
+I've also added `ParallelExecutor` and `ResultsGroup` classes that abstract away parallel execution and results aggregation, respectively. And finally, I added `multiprocessing` and `dask`/`dask.distributed` backends that reportedly speed up the analysis!
+
+## The current state.
+Currently, changes to the `AnalysisBase` are almost finalized. One thing that holds it back is some CI/CD issues causing tests to timeout, but all `AnalysisBase`-related tests run both locally and on CI/CD system.
+
+## What's left to do.
+Optional thing suggested within my proposal was to actually add parallelization to the subclasses, and update tests accordingly. It turned out to be a tedious task, but finally the mechanism is finalized and described in `MDAnalysisTests/analysis/conftest.py`. It automatically generates each subclass fixtures for testing, and updating tests accordingly is fairly simple yet tedious. After updating all the tests, the library will be fully equipped with the parallel execution mechanisms for those classes that allow it.
+
+
+## What code got merged (or not) upstream.
```

Comment:
@marinegor - Could you please provide some text to address this?

Reply:
>Could you please provide some text to address this?

Replace the paragraph with this:

---
Optional thing suggested within my proposal was to actually add parallelization to the subclasses, and update tests accordingly. It turned out to be a tedious task, but finally the mechanism is finalized and described in `MDAnalysisTests/analysis/conftest.py`, generating appropriate fixtures for testing subclasses with all supported (for the subclass) and installed (in the environment) backends.

Since it is a big change for the library, essentially changing the API for "analysis" part of the MDAnalysis, it requires a lot of time to merge into the `develop` branch, and wasn't merged within the GSoC timeframe. But the feature is expected by version 3.0 in the core library🚀
---

@RMeli am I correct about 3.0 as well?

Reply:
Hopefully we can merge it before. =) But it's definitely a milestone for 3.0!

Reply:
Thank you! I've added it in and linked to the 3.0 roadmap where you say 'version 3.0'.